src/diagnostics/error-guaranteed.md

```diff
@@ -0,0 +1,36 @@
+# `ErrorGuaranteed`
+
+The previous sections have been about the error message that a user of the
+compiler sees. But emitting an error can also have a second important side
+effect within the compiler source code: it generates an
+[`ErrorGuaranteed`][errorguar].
+
+`ErrorGuaranteed` is a zero-sized type that is unconstructable outside of the
+[`rustc_errors`][rerrors] crate. It is generated whenever an error is reported
+to the user, so that if your compiler code ever encounters a value of type
+`ErrorGuaranteed`, the compilation is _statically guaranteed to fail_. This is
+useful for avoiding unsoundness bugs because you can statically check that an
+error code path leads to a failure.
+
+There are some important considerations about the usage of `ErrorGuaranteed`:
+
+* It does _not_ convey information about the _kind_ of error. For example, the
+  error may be due (indirectly) to an ICE. Thus, you should not rely on
```

Comment:
I don't understand quite what this means! If there was an ICE, then how can we still be executing?

Reply:
I think I meant a delay span bug or other compiler error. Tbh, this is a bit cold in the cache for me...

Reply:
Understandable :)

Reply:
If you want to merge it, I'm ok with that, something is better than nothing!

Reply:
```suggestion
  error may be due (indirectly) to a `delay_span_bug` or other compiler error.
  Thus, you should not rely on
```